Basic_viewer/doc/Basic_viewer/CGAL/Qt/Basic_viewer_qt.h

```diff
@@ -0,0 +1,256 @@
+
+  /// returns the graphics scene of the viewer.
+  Graphics_scene<BufferType>& graphics_scene();
+
+  /// returns the graphics scene of the viewer, const version.
+  const Graphics_scene<BufferType>& graphics_scene() const;
+
+  /// reverses all normals of vertices and faces.
+  void reverse_all_normals();
+
+  /// returns `true` if the data structure in in 2D, i.e., lies on a plane.
+  bool two_dimensional() const;
+
+  /// draws the viewer without recomputing all internal buffers.
+  virtual void draw();
+
+  /// redraws the viewer, i.e., recompute all internal buffers and update the window.
+  virtual void redraw();
+
+  /// Function called when a key is pressed. Users can define their own function in order
+  /// to add specific behavior.
+  std::function<bool(QKeyEvent *, CGAL::Basic_viewer_qt<BufferType> *)> on_key_pressed;
+};
+
+
+//------------------------------------------------------------------------------
+/*!
+ \ingroup PkgBasicViewerClasses
+
+The class `QApplication_and_basic_viewer` regroups a `Basic_viewer_qt` and Qt `QApplication`. The `QApplication` is created in the constructor, but started by the `run()` method. This allows for example users to modify the `on_key_pressed` method of the `Basic_viewer_qt` to define their own behavior. This class requires `CGAL_Qt5`, and is only available if the macro `CGAL_USE_BASIC_VIEWER` is defined. Linking with the cmake target `CGAL::CGAL_Basic_viewer` will link with `CGAL_Qt5` and add the definition `CGAL_USE_BASIC_VIEWER`.
+
+\tparam BufferType the number type used for point coordinates: `float` by default.
+
+*/
+template <typename BufferType=float>
+class QApplication_and_basic_viewer
+{
+public:
+  /// Constructor given a `Graphics_scene` and possibly a title.
+  QApplication_and_basic_viewer(CGAL::Graphics_scene<BufferType>& buffer,
+                                const char* title="CGAL Basic Viewer");
+
+  /// runs the `QApplication`, i.e., open the Qt window. A call to this method is blocking, that is the program continues as soon as the user closes the window.
+  void run();
+
+  /// returns a reference to the `Basic_viewer_qt` associated with this.
+  Basic_viewer_qt<BufferType>& basic_viewer();
+};
+
+//------------------------------------------------------------------------------
+/*!
+  opens a new window and draws the given `Graphics_scene` (which must have been filled before). `title` will be the title of the window.  A call to this method is blocking, that is the program continues as soon as the user closes the window. This function requires `CGAL_Qt5`, and is only available if the macro `CGAL_USE_BASIC_VIEWER` is defined. Linking with the cmake target `CGAL::CGAL_Basic_viewer` will link with `CGAL_Qt5` and add the definition CGAL_USE_BASIC_VIEWER.
+*/
+template <typename BufferType=float>
+void draw_graphics_scene(Graphics_scene<BufferType>& graphic_buffer,
+                         const char *title="CGAL Basic Viewer")
+{}
```

Comment:
I do not get why the `graphic_buffer` (naming: buffer or scene?!) is taken by non-const reference. How can a "draw" function modify any of its inputs?

Reply:
it is now const.

Reply:
> it is now const.

Well now I will have a look at those `mutable` you introduced. :smile: